mitmproxy/proxy/protocol/websocket.py

```diff
@@ -81,27 +81,80 @@ def _handle_event(self, event, source_conn, other_conn, is_server):
+            return self._handle_connection_closed(event, source_conn, other_conn, is_server)
 
         # fail-safe for unhandled events
         return True
 
     def _handle_data_received(self, event, source_conn, other_conn, is_server):
+        fb = self.server_frame_buffer if is_server else self.client_frame_buffer
+        fb.append(event.data)
+
+        if event.message_finished:
+            original_chunk_sizes = [len(f) for f in fb]
+            message_type = 0x01 if isinstance(event, events.TextReceived) else 0x02
+            if message_type == 0x01:
+                payload = ''.join(fb)
+            else:
+                payload = b''.join(fb)
+            fb.clear()
+
+            websocket_message = WebSocketMessage(message_type, not is_server, payload)
+            length = len(websocket_message.content)
+            self.flow.messages.append(websocket_message)
+            self.channel.ask("websocket_message", self.flow)
+
+            if not self.flow.stream:
+                def get_chunk(payload):
+                    if len(payload) == length:
+                        # message has the same length, we can reuse the same sizes
+                        pos = 0
+                        for s in original_chunk_sizes:
+                            yield (payload[pos:pos + s], True if pos + s == length else False)
+                            pos += s
+                    else:
+                        # just re-chunk everything into 4kB frames
+                        # header len = 4 bytes without masking key and 8 bytes with masking key
+                        chunk_size = 4092 if is_server else 4088
+                        chunks = range(0, len(payload), chunk_size)
+                        for i in chunks:
+                            yield (payload[i:i + chunk_size], True if i + chunk_size >= len(payload) else False)
+
+                for chunk, final in get_chunk(websocket_message.content):
+                    self.connections[other_conn].send_data(chunk, final)
+
+                other_conn.send(self.connections[other_conn].bytes_to_send())
+
+            else:
+                self.connections[other_conn].send_data(event.data, event.message_finished)
+                other_conn.send(self.connections[other_conn].bytes_to_send())
+
+        elif self.flow.stream:
+            self.connections[other_conn].send_data(event.data, event.message_finished)
+            other_conn.send(self.connections[other_conn].bytes_to_send())
+
         return True
 
     def _handle_ping_received(self, event, source_conn, other_conn, is_server):
         # PING is automatically answered with a PONG by wsproto
-        # TODO: log this PING and its payload
-        self.connections[other_conn].ping(event.payload)
+        self.connections[other_conn].ping()
         other_conn.send(self.connections[other_conn].bytes_to_send())
+        source_conn.send(self.connections[source_conn].bytes_to_send())
```

Comment:
Once the we get a PingReceived event a PONG frame will eventually be sent to the sender as it will be in the buffer but the pong is not being sent in L259.
Receive_bytes only adds bytes to the buffer so the pong will not be present before the PingReceived event is generated in the for loop